drivers/sensor/ad7156/ad7156_trigger.c

```diff
@@ -0,0 +1,173 @@
+/*
+ * Copyright (c) 2023 Sensorfy B.V.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/device.h>
+#include <zephyr/drivers/gpio.h>
+#include <zephyr/drivers/i2c.h>
+#include <zephyr/sys/util.h>
+#include <zephyr/kernel.h>
+#include <zephyr/drivers/sensor.h>
+
+#include "ad7156.h"
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_DECLARE(AD7156, CONFIG_SENSOR_LOG_LEVEL);
+
+static void setup_int(const struct device *dev,
+		      bool enable)
+{
+	const struct ad7156_dev_config *cfg = dev->config;
+	gpio_flags_t flags = enable
+		? GPIO_INT_EDGE_TO_ACTIVE
+		: GPIO_INT_DISABLE;
+
+	gpio_pin_interrupt_configure_dt(&cfg->int_gpio, flags);
+}
+
+static void handle_int(const struct device *dev)
+{
+	struct ad7156_data *drv_data = dev->data;
+
+	setup_int(dev, false);
+
+#if defined(CONFIG_AD7156_TRIGGER_OWN_THREAD)
+	k_sem_give(&drv_data->gpio_sem);
+#elif defined(CONFIG_AD7156_TRIGGER_GLOBAL_THREAD)
+	k_work_submit(&drv_data->work);
+#endif
+}
+
+static void process_int(const struct device *dev)
+{
+	struct ad7156_data *drv_data = dev->data;
+	const struct ad7156_dev_config *cfg = dev->config;
+	uint8_t status;
+
+	/* Clear the status */
+	if (i2c_reg_read_byte_dt(&cfg->i2c,
+				 AD7156_REG_STATUS, &status) < 0) {
+		return;
+	}
+
+	if (drv_data->th_handler != NULL) {
+		drv_data->th_handler(dev, drv_data->th_trigger);
+	}
+
+	setup_int(dev, true);
+
+	/* Check for pin that asserted while we were offline */
+	int pv = gpio_pin_get_dt(&cfg->int_gpio);
+
+	if (pv > 0) {
+		handle_int(dev);
```

Comment:
all this `handle` and `process` and `setup` is a bit of spaghetti to me. Some better naming might help, but I see this is copied from other drivers so perhaps there's value in staying consistent as well?

Reply:
I agree and I don't want to make this anymore difficult when I push this to Zephyr.